### PR TITLE
Add upstream to forked repositories

### LINKED ISSUE #2364: Add an upstream to forked repositories
### Description

Currently, cloning a forked repository adds an `origin` remote linking to the fork on GitHub. However, there is no way to fetch changes from the parent, because no remote is added for the parent.

### Version

**GitHub Desktop version:** 0.7.0

**OS version:** macOS 10.12.3 (16D32)

### Steps to Reproduce

1. Fork a repo on GitHub
1. Clone the fork with GitHub Desktop
1. Run `git remote` on the command line

**Expected behavior:** Two remotes are present, with `origin` linking to the fork and `upstream` or *`username-of-parent-repo`* linking to the parent repo

**Actual behavior:** Only the `origin` remote is present

**Reproduces how often:** Always

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #2364

1. When a repository is selected, check if it's a fork.
1. If it's a fork, check if it has an `upstream` remote.
1. If it doesn't, add it.
1. If it does, check if it equals the remote we would have added.
1. If it doesn't, ask the user if they wanna update it.

### CODE REVIEW

**Comment by joshaber** [comment]
I could use some 🤔 on this dialog. On the one hand, I wanna give them as much context as possible. On the other hand, this is pretty overwhelming as-is.

screen shot 2017-10-30 at 3 53 22 pm

@desktop/core thoughts?

**Comment by joshaber** [comment]
This is ready 🙏 Though as I said above I'd like feedback specifically about the copy on that dialog ☝️

**Comment by lee-dohm** [comment]
@joshaber I would use something like this for the dialog body text:

> The repository `react-native-interactable` is a fork of `wix/react-native-interactable` but its `upstream` remote doesn't appear to point to `wix/react-native-interactable` as we would expect. Would you like us to set the `upstream` remote to the expected value for you?
>
> * Current: `https://github.com/wix/react-native-interactablasde.git`
> * Expected: `https://github.com/wix/react-native-interactable.git`

Also, if they don't choose to update the `upstream` remote, is there anything that might not work as expected?

**Comment by Haacked** [comment]
I was just going to suggest what @lee-dohm did. For non-github repos, we could show the full URL. For GitHub repos, we could tooltip the name/owner to show the full URL.

Also, we may not need to repeat the fork name/owner pair in the message.

```
The repository react-native-interactable is a fork of wix/react-native-interactable but its upstream remote points elsewhere. Would you like us to set the upstream remote to the expected value for you?

Current: https://github.com/wix/react-native-interactablasde.git
Expected: https://github.com/wix/react-native-interactable.git
```

**Comment by joshaber** [comment]
Thanks @lee-dohm @Haacked 💖 

Updated:

screen shot 2017-10-31 at 8 52 32 am

File: app/src/ui/upstream-already-exists/upstream-already-exists.tsx [hunk 1]
```diff
@@ -52,14 +52,17 @@ export class UpstreamAlreadyExists extends React.Component<
...
-            <Ref>{UpstreamRemoteName}</Ref> remote is <Ref>{existingURL}</Ref>{' '}
-            instead of <Ref>{replacementURL}</Ref>
+            The repository <Ref>{name}</Ref> is a fork of{' '}
+            <Ref>{parentName}</Ref>, but its <Ref>{UpstreamRemoteName}</Ref>{' '}
+            remote points elsewhere.
           </p>
           <p>
-            Would you like to update the <Ref>{UpstreamRemoteName}</Ref> remote
-            to use the expected URL?
+            • Current: <Ref>{existingURL}</Ref>
```
Comment (j-f1) [inline app/src/ui/upstream-already-exists/upstream-already-exists.tsx]: Shouldn’t this use a `<ul>`?

```jsx
<ul>
  <li>Current: <Ref>{existingURL}</Ref></li>
  <li>Expected: <Ref>{replacementURL}</Ref></li>
</ul>
```

Comment (joshaber) [reply app/src/ui/upstream-already-exists/upstream-already-exists.tsx]: Yup, good catch 👍 I forgot to update it after yolo'ing the message.

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -2874,4 +2873,42 @@ export class AppStore {
...
+
+    return Promise.resolve()
+  }
+
+  private getIgnoreExistingUpstreamRemote(
+    repository: Repository
+  ): Promise<boolean> {
+    const key = this.getIgnoreExistingUpstreamRemoteKey(repository)
+    const value = localStorage.getItem(key)
+    return Promise.resolve(value ? true : false)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/app-store.ts]: Can we be explicit in this check? As it stands, if `value` is equal to '1' or '0', the function would return true. I'm assuming the intent is to only return true when `value == '1'`.

File: app/src/lib/stores/git-store.ts [hunk 3]
```diff
@@ -689,6 +705,87 @@ export class GitStore {
...
+      addRemote(this.repository, UpstreamRemoteName, url)
+    )
+    this._upstream = { name: UpstreamRemoteName, url }
+  }
+
+  private findUpstreamRemote(
+    parent: GitHubRepository,
+    remotes: ReadonlyArray<IRemote>
+  ): IRemote | null {
+    const upstream = remotes.find(r => r.name === UpstreamRemoteName)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/git-store.ts]: What if the user named their remote something other than upstream?

Comment (joshaber) [reply app/src/lib/stores/git-store.ts]: The same as if they didn't have it: we'd add it.

**Comment by joshaber** [comment]
giphy-2

File: app/src/lib/stores/app-store.ts [hunk 4]
```diff
@@ -2899,7 +2899,7 @@ export class AppStore {
   ): Promise<boolean> {
     const key = this.getIgnoreExistingUpstreamRemoteKey(repository)
     const value = localStorage.getItem(key)
-    return Promise.resolve(value ? true : false)
+    return Promise.resolve(value === '1' ? true : false)
```
Comment (j-f1) [inline app/src/lib/stores/app-store.ts]: Couldn’t this just be `Promise.resolve(value === '1')`?

**Comment by j-f1** [comment]
@tooomm Can you open a new issue and fill out the template so we can better help you?

**Comment by tooomm** [comment]
sure --> https://github.com/desktop/desktop/issues/3252

**Comment by KirkMunro** [comment]
With this fix, will the repository properties dialog show both the origin and upstream repositories?

Do you know when this fix will ship?

**Comment by iAmWillShepherd** [comment]
> With this fix, will the repository properties dialog show both the origin and upstream repositories?

@KirkMunro , no this PR deals only with allowing you to checkout a branch from a fork. We are planning on shipping sometime this month.

**Comment by shiftkey** [comment]
@KirkMunro please open a new issue if you would like to see the upstream remote URL in the repository settings dialog

**Pull request merged**